File: backend/routers/sync.py
from fastapi import APIRouter, Depends, HTTPException, Body
from sqlalchemy.orm import Session
from database import get_db
from models import Transaction, Account, Category, Merchant, Subscription, Setting
from schemas import SyncPushRequest, SyncPullResponse, TransactionBase, AccountBase, CategoryBase, MerchantBase, SubscriptionBase, SettingBase
import logging

router = APIRouter()
logger = logging.getLogger(__name__)

# Helper to merge data
def merge_items(db: Session, model_class, items_data):
    for item_data in items_data:
        # Pydantic model to dict
        data_dict = item_data.model_dump()
        
        # Check if exists
        existing = db.query(model_class).filter(model_class.id == data_dict['id']).first()
        
        if existing:
            # Update existing
            for key, value in data_dict.items():
                setattr(existing, key, value)
        else:
            # Create new
            new_item = model_class(**data_dict)
            db.add(new_item)

@router.post("/sync/push")
async def sync_push(
    payload: SyncPushRequest,
    db: Session = Depends(get_db)
):
    """
    Receive data from frontend and merge into backend SQLite.
    Simplistic strategy: "Last write wins" (Frontend pushes state).
    """
    try:
        # Merge all entities
        # Note: Order matters due to foreign keys if we enforced them strictly,
        # but for sync usually we just upsert all.
        
        # Settings
        for setting in payload.settings:
            data = setting.model_dump()
            existing = db.query(Setting).filter(Setting.key == data['key']).first()
            if existing:
                existing.value = data['value']
                existing.updatedAt = data['updatedAt']
            else:
                db.add(Setting(**data))
        
        # Delete deleted transactions
        if payload.deletedTransactionIds:
            db.query(Transaction).filter(Transaction.id.in_(payload.deletedTransactionIds)).delete(synchronize_session=False)

        # Merge other entities
        merge_items(db, Account, payload.accounts)
        merge_items(db, Category, payload.categories)
        merge_items(db, Merchant, payload.merchants)
        merge_items(db, Subscription, payload.subscriptions)
        
        # Merge Transactions with Hash Collision Handling
        for item_data in payload.transactions:
            data = item_data.model_dump()
            
            # 1. Check by ID (Update)
            existing_id = db.query(Transaction).filter(Transaction.id == data['id']).first()
            if existing_id:
                for key, value in data.items():
                    setattr(existing_id, key, value)
                continue
            
            # 2. Check by Hash (Collision / Re-upload)
            existing_hash = db.query(Transaction).filter(Transaction.transactionHash == data['transactionHash']).first()
            if existing_hash:
                # Collision found! The user likely re-uploaded.
                # We should replace the old record with the new one to match the Frontend's new ID.
                db.delete(existing_hash)
                db.flush() # Ensure delete happens before insert
            
            # 3. Insert New
            db.add(Transaction(**data))
        
        db.commit()
        return {"status": "success", "message": "Sync complete"}
    except Exception as e:
        db.rollback()
        logger.error(f"Sync failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/sync/pull", response_model=SyncPullResponse)
async def sync_pull(db: Session = Depends(get_db)):
    """
    Send entire backend state to frontend.
    """
    try:
        return {
            "transactions": db.query(Transaction).all(),
            "accounts": db.query(Account).all(),
            "categories": db.query(Category).all(),
            "merchants": db.query(Merchant).all(),
            "subscriptions": db.query(Subscription).all(),
            "settings": db.query(Setting).all()
        }
    except Exception as e:
        logger.error(f"Pull failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: src/lib/sync.ts
import { db } from './db';
import type { Transaction, Account, Category, Merchant, Subscription, Setting } from '@/types/database';

const API_BASE = 'http://localhost:8000/api';

/**
 * Pull data from Backend (SQLite) to Frontend (IndexedDB)
 * Should be called on app startup.
 */
export async function syncPull(): Promise<void> {
    try {
        const response = await fetch(`${API_BASE}/sync/pull`);
        if (!response.ok) throw new Error('Failed to fetch sync data');

        const data = await response.json();

        // Check for pending local deletions
        const deletedLogs = await db.deletedLog.toArray();
        const deletedTransactionIds = new Set(
            deletedLogs
                .filter(log => log.type === 'transaction')
                .map(log => log.id)
        );

        // Restore Transactions
        if (data.transactions && data.transactions.length > 0) {
            const transactions = data.transactions
                .filter((t: any) => !deletedTransactionIds.has(t.id)) // Filter out locally deleted items
                .map((t: any) => ({
                    ...t,
                    date: new Date(t.date),
                    createdAt: new Date(t.createdAt),
                    updatedAt: new Date(t.updatedAt),
                }));
            await db.transactions.bulkPut(transactions);
        }

        // Restore Accounts
        if (data.accounts && data.accounts.length > 0) {
            const accounts = data.accounts.map((a: any) => ({
                ...a,
                lastProcessedDate: a.lastProcessedDate ? new Date(a.lastProcessedDate) : null,
                createdAt: new Date(a.createdAt),
                updatedAt: new Date(a.updatedAt),
            }));
            await db.accounts.bulkPut(accounts);
        }

        // Restore Categories
        if (data.categories && data.categories.length > 0) {
            const categories = data.categories.map((c: any) => ({
                ...c,
                createdAt: new Date(c.createdAt),
                updatedAt: new Date(c.updatedAt),
            }));
            await db.categories.bulkPut(categories);
        }

        // Restore Merchants
        if (data.merchants && data.merchants.length > 0) {
            const merchants = data.merchants.map((m: any) => ({
                ...m,
                createdAt: new Date(m.createdAt),
                updatedAt: new Date(m.updatedAt),
            }));
            await db.merchants.bulkPut(merchants);
        }

        // Restore Subscriptions
        if (data.subscriptions && data.subscriptions.length > 0) {
            const subscriptions = data.subscriptions.map((s: any) => ({
                ...s,
                firstChargeDate: s.firstChargeDate ? new Date(s.firstChargeDate) : new Date(), // Fallback
                lastChargeDate: s.lastChargeDate ? new Date(s.lastChargeDate) : new Date(), // Fallback
                nextExpectedDate: s.nextExpectedDate ? new Date(s.nextExpectedDate) : null,
                createdAt: new Date(s.createdAt),
                updatedAt: new Date(s.updatedAt),
            }));
            await db.subscriptions.bulkPut(subscriptions);
        }

        // Restore Settings
        if (data.settings && data.settings.length > 0) {
            const settings = data.settings.map((s: any) => ({
                ...s,
                updatedAt: new Date(s.updatedAt),
            }));
            await db.settings.bulkPut(settings);
        }

        console.log('[Sync] Pull completed successfully.');
    } catch (err) {
        console.error('[Sync] Pull failed:', err);
        // Don't throw, just log. App can work offline/locally.
    }
}

/**
 * Push data from Frontend (IndexedDB) to Backend (SQLite)
 * Should be called after significant mutations.
 */
export async function syncPush(): Promise<void> {
    try {
        const [
            transactions,
            accounts,
            categories,
            merchants,
            subscriptions,
            settings,
            deletedLogs
        ] = await Promise.all([
            db.transactions.toArray(),
            db.accounts.toArray(),
            db.categories.toArray(),
            db.merchants.toArray(),
            db.subscriptions.toArray(),
            db.settings.toArray(),
            db.deletedLog.toArray(),
        ]);

        // Filter deleted IDs for transactions (can be extended)
        const deletedTransactionIds = deletedLogs
            .filter(log => log.type === 'transaction')
            .map(log => log.id);

        const payload = {
            transactions,
            accounts,
            categories,
            merchants,
            subscriptions,
            settings,
            deletedTransactionIds
        };

        const response = await fetch(`${API_BASE}/sync/push`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });

        if (!response.ok) throw new Error('Failed to push sync data');

        // On success, clear the deleted logs
        if (deletedLogs.length > 0) {
            await db.deletedLog.clear();
        }

        console.log('[Sync] Push completed successfully.');
    } catch (err) {
        console.error('[Sync] Push failed:', err);
    }
}
